File: src/main/kotlin/uk/ac/ic/matterialize/camera/FakeWebcam.kt
package uk.ac.ic.matterialize.camera

class FakeWebcam(private val device: String, private val width: Int, private val height: Int) {
    private val v4l2 = V4L2Lib()
    private var dev = -1

    fun start(): Int {
        dev = v4l2.open(device, width, height)
        return dev
    }

    fun stop() {
        v4l2.close(dev)
    }

    fun write(frame: ByteArray?): Boolean {
        return v4l2.writeFrame(dev, frame)
    }
}

File: src/files/camera/fake_webcam.cpp
#include "fake_webcam.hpp"
#include "../util/converter.hpp"

using namespace std;

void FakeWebcam::write(const cv::Mat &frame) {
  vector<uchar> converted = Converter::to_YUYV(frame);
  output->write((char *) &converted[0], converted.size());
}

void FakeWebcam::stop() {
  output->stop();
}

void FakeWebcam::start() {
  output->start();
}

cv::Size FakeWebcam::get_size() {
  return cv::Size(width, height);
}

FakeWebcam::FakeWebcam(int device, int width, int height)
    : device{device}, width{width}, height{height}, fps{30} {
  string device_str = "/dev/video" + to_string(device);

  V4L2DeviceParameters params(
      device_str.c_str(), V4L2_PIX_FMT_YUYV, width, height, fps);

  output = V4l2Output::create(params, V4l2Access::IOTYPE_READWRITE);
}
